apps/web/src/pages/workspace/index.tsx

```diff
@@ -1,23 +1,40 @@
+import { useDataCenter, useDataCenterWorkspace } from '@affine/store';
 import { useRouter } from 'next/router';
-import { useCallback, useEffect } from 'react';
+import { useMemo } from 'react';
 
 import { PageLoading } from '@/components/loading';
-import useEnsureWorkspace from '@/hooks/use-ensure-workspace';
-import { useGlobalState } from '@/store/app';
 
-export const WorkspaceIndex = () => {
+// refactor(himself65): move this to top level hook
+function useCurrentWorkspace() {
   const router = useRouter();
-  const currentWorkspace = useGlobalState(
-    useCallback(store => store.currentDataCenterWorkspace, [])
+  const dataCenter = useDataCenter();
+  const workspaceId =
+    typeof router.query.workspaceId === 'string'
+      ? router.query.workspaceId
+      : dataCenter.workspaces.at(0)?.id ?? null;
+  const workspace = useDataCenterWorkspace(workspaceId);
+  const notExist = useMemo(
+    () =>
+      workspaceId &&
+      dataCenter.workspaces.length &&
+      dataCenter.workspaces.findIndex(
+        meta => meta.id.toString() === workspaceId
+      ) === -1,
+    [dataCenter.workspaces, workspaceId]
   );
-  const { workspaceLoaded } = useEnsureWorkspace();
+  if (notExist) {
```

Comment:
better put side effects in useEffect in case it will trigger multiple times because of rerendering